### PR TITLE
rgw: S3 Bucket Policy Conditions IpAddress and NotIpAddress do not work

### PR DESCRIPTION
**Pull request by jgibson**
This fixes bugs in rgw's support for the bucket policy conditions IpAddress and NotIpAddress.

The only concern I have with the patch is getting the correct byte order for the ip addresses themselves on big endian architectures. (Although I don't know if ceph actually suppports big endian architectures.)

http://tracker.ceph.com/issues/20991

### CODE REVIEW

File: src/rgw/rgw_iam_policy.cc [hunk 1]
```diff
@@ -1091,27 +1091,27 @@ optional<MaskedIP> Condition::as_network(const string& s) {
...
-    m.addr |= Address(a.sin6_addr.s6_addr[7]) << 56;
-    m.addr |= Address(a.sin6_addr.s6_addr[8]) << 64;
-    m.addr |= Address(a.sin6_addr.s6_addr[9]) << 72;
-    m.addr |= Address(a.sin6_addr.s6_addr[10]) << 80;
-    m.addr |= Address(a.sin6_addr.s6_addr[11]) << 88;
-    m.addr |= Address(a.sin6_addr.s6_addr[12]) << 96;
-    m.addr |= Address(a.sin6_addr.s6_addr[13]) << 104;
-    m.addr |= Address(a.sin6_addr.s6_addr[14]) << 112;
-    m.addr |= Address(a.sin6_addr.s6_addr[15]) << 120;
+    m.addr |= Address(a.s6_addr[15]) << 0;
```
Comment (mattbenjamin) [inline src/rgw/rgw_iam_policy.cc]: ha!

**Comment by mattbenjamin** [comment]
(this was supposed to be a comment)

**Comment by smithfarm** [comment]
This PR is targeting luminous - is luminous still getting merged to master?

**Comment by smithfarm** [comment]
I believe this PR should be retargeted to master, and then cherry-pick to luminous after merge. Correct me if I'm wrong!

**Comment by theanalyst** [comment]
@jgibson can you rebase this PR against ceph master and retarget the PR against master... The process is that PRs should target master and then we backport them back to Luminous

**Comment by smithfarm** [comment]
make check failure:

```
The following tests FAILED:
	146 - unittest_rgw_iam_policy (Failed)
```

**Comment by jgibson** [comment]
Sure, I can rebase against master, @theanalyst. I targeted luminous because I was fixing a bug in an existing release and didn't read the instructions in [SubmittingPatches.rst](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst) closely enough.

Although first I'll try to figure out why the new IP address unit test failed (they were fine on my dev box). What build environment do you use?

**Comment by jgibson** [comment]
OK, I finally sorted out my mistake in the earlier pushes. It took longer than I would've liked, but on the upside the tests are more comprehensive now. Let me know if there's anything more that you need me to do to get this pull request merged.

**Comment by adamemerson** [comment]
Is this still relevant? I think we had another PR patching this that got merged a bit ago.

**Comment by jgibson** [comment]
I can't find the relevant PR, so I'm not sure. We've needed the IP address conditions for production and have just been running from a version of luminous compiled with those changes for a while now. I do see that now there are some conflicting files. I'll see if I can see what changed.

**Comment by jgibson** [comment]
OK, I took a look at master and this is definitely still relevant. At a minimum the current version is missing support for a custom remote address header, which is necessary for hosting ceph behind a reverse proxy and having the IP address conditions still be useful.

I'd also wager that master's NotIpAddress condition will not work properly when there are multiple addresses or ranges in the list.

In addition I didn't see any unit tests of the new functionality on master, whereas this PR has several. I can take a look at updating this PR to merge cleanly with the changes that have already been made soon.

**Comment by adamemerson** [comment]
All right, thank you. As soon as we can get this clean I'll try and get it merged in.

**Comment by jgibson** [comment]
OK, I've resolved the conflicts. The changes introduced in 2fb445b (and possibly b3118cab) did resolve most of the problems with IPv4 addresses.

The remaining changes in this PR fix IPv6, the string representation of the MaskedIPs, and support for NotIpAddress when using multiple values in the condition. In addition it adds support for evaluating against the request header specified rgw_remote_addr_param (typically used by reverse proxies). It also includes proper support for the X-Forwarded-For header when a request has passed through multiple remote proxies.

Note that I've only verified this last round of changes via the unit tests. A full test on our running system (running luminous under [Rook](https://github.com/rook/ceph/)) will have to wait until after New Year's.

**Comment by jgibson** [comment]
I did some initial unit tests for the luminous backport and it appears that you will have to cherry pick 650d30d, a7184ca8, 9787fe6b, and 2fb445b before you can apply the changes in this pull request to the luminous branch.

**Comment by adamemerson** [comment]
@yuriw Could you include this in your next QA run?

**Comment by adamemerson** [comment]
@yuriw Never mind! My QA run actually finished. Sorry for the false alarm.

**Comment by adamemerson** [comment]
http://pulpito.ceph.com:80/aemerson-2018-01-05_22:40:56-rgw-wip-aemerson-test-1701-distro-basic-smithi/

**Pull request merged**